### PR TITLE
Add span type inference

### PR DESCRIPTION
**Pull request by jjonescz**
Test plan: https://github.com/dotnet/roslyn/issues/73445
Corresponding speclet update: https://github.com/dotnet/csharplang/pull/8333

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit3/FirstClassSpanTests.cs [hunk 1]
```diff
@@ -4361,26 +4359,31 @@ static class C
...
+            Diagnostic(ErrorCode.ERR_NoSuchMemberOrExtension, "M").WithArguments("int[]", "M").WithLocation(11, 28));
 
-        // PROTOTYPE: Some of these need type inference to work.
         var expectedDiagnostics = new[]
         {
-            // (4,10): error CS8917: The delegate type could not be inferred.
+            // (4,12): error CS0123: No overload for 'M' matches delegate 'Func<int, int>'
             // var d1 = a.M;
-            Diagnostic(ErrorCode.ERR_CannotInferDelegateType, "a.M").WithLocation(4, 10),
+            Diagnostic(ErrorCode.ERR_MethDelegateMismatch, "M").WithArguments("M", "System.Func<int, int>").WithLocation(4, 12),
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Emit3/FirstClassSpanTests.cs:4374]: Note that type inference successfully infers the delegate type to be `Func<int, int>` but then binding fails (because the span conversion doesn't exist for extension method receiver). I'm not sure if this is a problem, I think this could happen for some existing scenarios as well.

**Comment by jjonescz** [comment]
@cston @333fred for reviews, thanks

File: src/Compilers/CSharp/Test/Emit3/FirstClassSpanTests.cs [hunk 2]
```diff
@@ -7408,4 +7543,282 @@ static class C
...
+    }
+
+    [Fact]
+    public void TypeInference_ReadOnlySpan_ReadOnlySpan_02()
+    {
+        var source = """
+            using System;
+            class C
+            {
+                void M(ReadOnlySpan<string> a, object b) => M1(a, b);
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit3/FirstClassSpanTests.cs:7753]: Let's test a scenario like this with a non-variant non-array conversion, like `ROS<int>` and `object`/vice versa; I expect the former to fail, and the latter to succeed.

**Review by 333fred — COMMENTED** [review]
Mostly LGTM. Couple of minor comments.

File: src/Compilers/CSharp/Test/Emit3/FirstClassSpanTests.cs [hunk 3]
```diff
@@ -279,6 +279,125 @@ static class E
...
+            static class C
+            {
+                public static void R<T>(IEnumerable<T> e) => Console.Write(1);
+                public static void R<T>(Span<T> s) => Console.Write(2);
+            }
+            """;
+        var comp = CreateCompilationWithSpanAndMemoryExtensions(source, parseOptions: TestOptions.Regular13);
+        CompileAndVerify(comp, expectedOutput: "1").VerifyDiagnostics();
+
+        var expectedOutput = ExecutionConditionUtil.IsCoreClr ? "3" : "2";
```
[the remark this answers is not in the document]
Comment (cston) [reply src/Compilers/CSharp/Test/Emit3/FirstClassSpanTests.cs:362]: Consider reporting the exception type rather than "3".

File: src/Compilers/CSharp/Test/Emit3/FirstClassSpanTests.cs [hunk 4]
```diff
@@ -279,6 +279,125 @@ static class E
...
+
+        comp = CreateCompilationWithSpanAndMemoryExtensions(source, parseOptions: TestOptions.RegularNext);
+        CompileAndVerify(comp, expectedOutput: expectedOutput).VerifyDiagnostics();
+
+        comp = CreateCompilationWithSpanAndMemoryExtensions(source);
+        CompileAndVerify(comp, expectedOutput: expectedOutput).VerifyDiagnostics();
+    }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit3/FirstClassSpanTests.cs:399]: >} [](http://example.com/codeflow?start=4&length=1)

Are we testing similar cases to these new tests, but with `IEnumerable<T>` and `ReadOnlySpan<T>`?

File: src/Compilers/CSharp/Test/Emit3/FirstClassSpanTests.cs [hunk 5]
```diff
@@ -7408,4 +7543,282 @@ static class C
...
+            {
+                void M(Span<string[]> a, object b) => M1(a, b);
+                void M1<T>(Span<T[]> x, T y) { }
+            }
+            """;
+        CreateCompilationWithSpanAndMemoryExtensions(source).VerifyDiagnostics(
+            // (4,43): error CS0411: The type arguments for method 'C.M1<T>(Span<T[]>, T)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+            //     void M(Span<string[]> a, object b) => M1(a, b);
+            Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "M1").WithArguments("C.M1<T>(System.Span<T[]>, T)").WithLocation(4, 43));
+    }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit3/FirstClassSpanTests.cs:7620]: >} [](http://example.com/codeflow?start=4&length=1)

Consider testing with `M(Span<object[]> a, string b) => M1(a, b)` as well.

File: src/Compilers/CSharp/Test/Emit3/FirstClassSpanTests.cs [hunk 6]
```diff
@@ -7408,4 +7543,282 @@ static class C
...
+            using System;
+            class C
+            {
+                void M(object[] a, string b) => M1(a, b);
+                void M1<T>(ReadOnlySpan<T> x, T y) { }
+            }
+            """;
+        var comp = CreateCompilationWithSpanAndMemoryExtensions(source).VerifyDiagnostics();
+        AssertEx.Equal("C.M1<System.Object>", DisplayInvokedMethodTypeArguments(comp));
+    }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit3/FirstClassSpanTests.cs:7729]: >} [](http://example.com/codeflow?start=4&length=1)

Please test:
```
void M2(int[] a, object b) => M1(a, b);
void M3(object[] a, int b) => M1(a, b);
```

File: src/Compilers/CSharp/Test/Emit3/FirstClassSpanTests.cs [hunk 7]
```diff
@@ -7408,4 +7543,282 @@ static class C
...
+            {
+                void M(ReadOnlySpan<int> a) => M1(a);
+                void M1<T>(Span<T> x) { }
+            }
+            """;
+        CreateCompilationWithSpanAndMemoryExtensions(source).VerifyDiagnostics(
+            // (4,36): error CS0411: The type arguments for method 'C.M1<T>(Span<T>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+            //     void M(ReadOnlySpan<int> a) => M1(a);
+            Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "M1").WithArguments("C.M1<T>(System.Span<T>)").WithLocation(4, 36));
+    }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit3/FirstClassSpanTests.cs:7823]: >} [](http://example.com/codeflow?start=4&length=1)

Are we testing inferring `ReadOnlySpan<T>` from call with `Span<U>`?
```csharp
void M(Span<string> a, object b) => M1(a, b);
void M1<T>(ReadOnlySpan<T> a, T b) { }
```

File: src/Features/CSharp/Portable/CodeRefactorings/UseRecursivePatterns/UseRecursivePatternsCodeRefactoringProvider.cs [hunk 8]
```diff
@@ -600,7 +600,7 @@ static bool CanConvertToSubpattern(IdentifierNameSyntax name, SemanticModel mode
 
         // Process all nodes to refactor in reverse to ensure nested nodes
         // are processed before the outer nodes to refactor.
-        foreach (var originalNode in nodes.Reverse())
+        foreach (var originalNode in Enumerable.Reverse(nodes))
```
Comment (cston) [inline src/Features/CSharp/Portable/CodeRefactorings/UseRecursivePatterns/UseRecursivePatternsCodeRefactoringProvider.cs:603]: >Enumerable.Reverse(nodes) [](http://example.com/codeflow?start=37&length=25)

Why was this change needed? Are we now binding `nodes.Reverse()` to `MemoryExtensions.Reverse<T>(Span<T>)`? And if so, is that an issue? (It looks like the array that is created from `var nodes = ... ToArray();` should not include array variance.)

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 9]
```diff
@@ -105,14 +105,10 @@ private static string Concat(string container, string name)
                 }
             }
             """;
-        internal const string s_collectionExtensionsWithSpan = s_collectionExtensions +
+        internal const string s_collectionExtensionsWithReadOnlySpan = s_collectionExtensions +
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:108]: >s_collectionExtensionsWithReadOnlySpan [](http://example.com/codeflow?start=30&length=38)

Alternatively, could we leave the name unchanged, and simply remove the `Report<T>(this in Span<T>)` overload? Would that result in many additional test changes?

File: src/Compilers/CSharp/Test/Emit3/FirstClassSpanTests.cs (hunk 7 above)
Comment (cston) [inline src/Compilers/CSharp/Test/Emit3/FirstClassSpanTests.cs:7823]: >} [](http://example.com/codeflow?start=4&length=1)

Consider testing variance when the array and span are the type arguments. For instance, with the following, for the calls with `I<object[]>` and `IOut<object[]>` I'd expect type inference to succeed but the subsequent argument conversion to fail, and for the call with `IIn<object[]>`, I'd expect type inference to fail.
```csharp
interface I<T> where T : allows ref struct { }
class A
{
    static void M(I<object[]> a) => M1(a);
    static void M1<T>(I<ReadOnlySpan<T>> x) { }
}
interface IOut<out T> where T : allows ref struct { }
class B
{
    static void M(IOut<object[]> a) => M1(a);
    static void M1<T>(IOut<ReadOnlySpan<T>> x) { }
}
interface IIn<in T> where T : allows ref struct { }
class C
{
    static void M(IIn<object[]> a) => M1(a);
    static void M1<T>(IIn<ReadOnlySpan<T>> x) { }
}
```

File: src/Features/CSharp/Portable/CodeRefactorings/UseRecursivePatterns/UseRecursivePatternsCodeRefactoringProvider.cs (hunk 8 above)
Comment (jjonescz) [reply src/Features/CSharp/Portable/CodeRefactorings/UseRecursivePatterns/UseRecursivePatternsCodeRefactoringProvider.cs:603]: > Are we now binding `nodes.Reverse()` to `MemoryExtensions.Reverse<T>(Span<T>)`?

Yes.

> And if so, is that an issue?

Yes, because `MemoryExtensions.Reverse` returns `void`.

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs (hunk 9 above)
Comment (jjonescz) [reply src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:108]: Not many additional changes, I can do that.

**Review by cston — APPROVED** [review]
